src/wallet/wallet.cpp

```diff
@@ -2739,8 +2747,16 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
                     return false;
                 }
 
-                LearnRelatedScripts(vchPubKey, g_change_type);
-                scriptChange = GetScriptForDestination(GetDestinationForKey(vchPubKey, g_change_type));
+                // If -changetype is not specified and g_address_type is not legacy,
+                // and any destination is P2WPKH or P2WSH, use P2WPKH for the change
+                // output.
+                OutputType change_type =
```

Comment:
IMO should avoid long expressions. Suggestion:
```cpp
OutputType change_type = g_change_type;
if (change_type == OUTPUT_TYPE_NONE) {
    if (any_destination_native_segwit && g_address_type != OUTPUT_TYPE_LEGAC) {
        change_type = OUTPUT_TYPE_BECH32;
    } else {
        change_type = g_address_type;
    }
}
```

Also, `CreateTransaction` is already a rather long function, could you factor out this to a new function? Then add tests for that in `src/wallet/test/wallet_tests.cpp`.

Reply:
Switched to your version.

I extracted the change type stuff to a new function. I made a note to add tests later.